Source/WebCore/platform/PublicSuffixStore.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "PublicSuffixStore.h"
+
+#include <wtf/CrossThreadCopier.h>
+#include <wtf/NeverDestroyed.h>
+
+namespace WebCore {
+
+PublicSuffixStore& PublicSuffixStore::singleton()
+{
+    static LazyNeverDestroyed<PublicSuffixStore> store;
+    static std::once_flag flag;
+    std::call_once(flag, [&] {
+        store.construct();
+    });
+    return store.get();
+}
+
+void PublicSuffixStore::clearHostTopPrivatelyControlledDomainCache()
+{
+    Locker locker { m_HostTopPrivatelyControlledDomainCacheLock };
+    m_hostTopPrivatelyControlledDomainCache.clear();
+}
+
+void PublicSuffixStore::enablePublicSuffixCache()
+{
+    Locker locker { m_publicSuffixCacheLock };
+    ASSERT(!m_publicSuffixCache);
+
+    m_publicSuffixCache = HashSet<String, ASCIICaseInsensitiveHash> { };
+}
+
+void PublicSuffixStore::addPublicSuffix(const String& publicSuffix)
+{
+    if (publicSuffix.isEmpty())
+        return;
+
+    Locker locker { m_publicSuffixCacheLock };
+
+    if (m_publicSuffixCache)
+        m_publicSuffixCache->add(crossThreadCopy(publicSuffix));
```

Comment:
Should we check if `publicSuffix` is valid with `platformIsPublicSuffix` before adding?

Reply:
Actually, I think the answer is no, since that would make a call to the CFNetwork function we are trying to avoid calling in the WebContent process.

Reply:
Yes, the added public suffix should be truth.